### PR TITLE
Fix: consistent behavior for hash methods

### PR DESCRIPTION
**Pull request by Wi1l-B0t**
# Description

The CrytoLib has several hash contract mehod, but different hash methods have different semantics if argument is null:
```
Sha256: ArgumentNullException if argument is null
RIPEMD160: NullReferenceException if argument is null
Murmur32 and Keccak256: OK if argument is null
```

This fix reduces restrictions rather than increasing for these methods.
After HF_Faun: All OK if argument is null

Fix: consistent behavior for hash methods. #4302

## Type of change

- [ ] Optimization (the change is only an optimization)
- [ ] Style (the change is only a code style for better maintenance or standard purpose)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [x] Unit Testing
- [ ] Run Application
- [ ] Local Computer Tests
- [ ] No Testing

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/Neo/SmartContract/Native/CryptoLib.cs [hunk 1]
```diff
@@ -60,22 +60,32 @@ internal CryptoLib() : base() { }
...
+        /// <param name="engine">The engine that is executing the method.</param>
         /// <param name="data">The input to compute the hash code for.</param>
         /// <returns>The computed hash code.</returns>
+        /// <exception cref="ArgumentNullException">Thrown if data is null before HF_Faun.</exception>
         [ContractMethod(CpuFee = 1 << 15, Name = "ripemd160")]
-        public static byte[] RIPEMD160(byte[] data)
+        public static byte[] RIPEMD160(ApplicationEngine engine, byte[] data)
         {
+            // Before HF_Faun, data cannot be null for compatibility
+            if (!engine.IsHardforkEnabled(Hardfork.HF_Faun)) ArgumentNullException.ThrowIfNull(data);
```
Comment (Wi1l-B0t) [inline src/Neo/SmartContract/Native/CryptoLib.cs:71]: The type of exception has changed.
`NullReferenceException` -> `ArgumentNullException`
Is this OK?

Comment (shargon) [reply src/Neo/SmartContract/Native/CryptoLib.cs:71]: Yes, it its

File: tests/Neo.UnitTests/SmartContract/Native/UT_CryptoLib.cs [hunk 2]
```diff
@@ -528,89 +528,64 @@ public void TestBls12381ScalarMul_Compat()
...
-            // Arrange
-            byte[] inputData = "Hello, World!"u8.ToArray();
-            string expectedHashHex = "acaf3289d7b601cbd114fb36c4d29c85bbfd5e133f14cb355c3fd8d99367964f";
-
-            // Act
-            byte[] outputData = CryptoLib.Keccak256(inputData);
-            string outputHashHex = Hex.ToHexString(outputData);
-
-            // Assert
+            var inputData = "Hello, World!"u8.ToArray();
```
Comment (shargon) [inline tests/Neo.UnitTests/SmartContract/Native/UT_CryptoLib.cs:531]: Could you add a test for each method where we can see that null return the same exception?

Comment (Wi1l-B0t) [reply tests/Neo.UnitTests/SmartContract/Native/UT_CryptoLib.cs:531]: > Could you add a test for each method where we can see that null return the same exception?

Always OK for each method even if data is null now, and tests added.

File: src/Neo/SmartContract/Native/CryptoLib.cs (hunk 1 above)
Comment (Jim8y) [reply src/Neo/SmartContract/Native/CryptoLib.cs:71]: exception type does not matter for vm

**Review by Jim8y — APPROVED** [review]

File: tests/Neo.UnitTests/SmartContract/Native/UT_CryptoLib.cs (hunk 2 above)
Comment (shargon) [reply tests/Neo.UnitTests/SmartContract/Native/UT_CryptoLib.cs:531]: I only see the Keccak256 null test

Comment (Wi1l-B0t) [reply tests/Neo.UnitTests/SmartContract/Native/UT_CryptoLib.cs:531]: > I only see the Keccak256 null test

`TestSha256` and `TestRIPMD160`.

Comment (shargon) [reply tests/Neo.UnitTests/SmartContract/Native/UT_CryptoLib.cs:531]: > TestRIPMD160

```csharp
public void TestRIPEMD160()
{
    ReadOnlySpan<byte> value = Encoding.ASCII.GetBytes("hello world");
    var result = value.RIPEMD160();
    Assert.AreEqual("98c615784ccb5fe5936fbc0cbe9dfdb408d92f0f", result.ToHexString());
}
```

where is the null?

File: src/Neo/SmartContract/Native/CryptoLib.cs (hunk 1 above)
Comment (erikzhang) [reply src/Neo/SmartContract/Native/CryptoLib.cs:71]: Why do we need hardfork here?

Comment (shargon) [reply src/Neo/SmartContract/Native/CryptoLib.cs:71]: We need to ensure that the state it's the same, otherwise a special script can transfer neo, and after the upgrade the logic fault the transaction. Any logic changes inside the VM require a hardfork or people can use it to scam

Comment (erikzhang) [reply src/Neo/SmartContract/Native/CryptoLib.cs:71]: Both `NullReferenceException` and `ArgumentNullException` fault the transaction. I think it doesn't matter.

Comment (erikzhang) [reply src/Neo/SmartContract/Native/CryptoLib.cs:71]: Okay. I find the problem. Before `HF_Faun`, we don't allow `null`. Now, we allow `null`.

Then, the question becomes, why do we allow `null` for a hash function?

Comment (shargon) [reply src/Neo/SmartContract/Native/CryptoLib.cs:71]: We can throw if null

Comment (erikzhang) [reply src/Neo/SmartContract/Native/CryptoLib.cs:71]: I mean, we don't need to fork here. We should always throw.

Comment (Wi1l-B0t) [reply src/Neo/SmartContract/Native/CryptoLib.cs:71]: > Okay. I find the problem. Before `HF_Faun`, we don't allow `null`. Now, we allow `null`.
> 
> Then, the question becomes, why do we allow `null` for a hash function?

Because some methods allow null and  others not allow.
If neo need to unify behavior, should choose the behavior  with fewer restrictions than before, rather than adding more restrictions.

Comment (Wi1l-B0t) [reply src/Neo/SmartContract/Native/CryptoLib.cs:71]: > > Okay. I find the problem. Before `HF_Faun`, we don't allow `null`. Now, we allow `null`.
> > Then, the question becomes, why do we allow `null` for a hash function?
> 
> Because some methods allow null and others not allow. If neo need to unify behavior, should choose the behavior with fewer restrictions than before, rather than adding more restrictions.

In this way,  previous usage doesn't have to change.
This is a common way to change API behavior;

Comment (Wi1l-B0t) [reply src/Neo/SmartContract/Native/CryptoLib.cs:71]: > I mean, we don't need to fork here. We should always throw.

Need a harfork to keep the same behaviour before HF_Faun.
